### PR TITLE
fix(upgrade): skip to check engine upgrade path

### PR DESCRIPTION
**Pull request by mantissahz**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#7539

#### What this PR does / why we need it:

Skip to check the engine upgrade path if the upgrade version check flag is disabled.

### CODE REVIEW

File: upgrade/util/util.go [hunk 1]
```diff
@@ -302,7 +302,7 @@ func checkEngineUpgradePath(namespace string, lhClient lhclientset.Interface, ne
 		return err
 	}
 
-	if len(engineImageVersions) == 0 {
+	if len(engineImageVersions) == 0 || !enableUpgradeVersionCheck {
```
Comment (innobead) [inline upgrade/util/util.go:305]: You can skip `!enableUpgradeVersionCheck` at the beginning of the method, but do it after querying the engine images.

**Review by innobead — CHANGES_REQUESTED** [review]
Can we do this?

Move the following code from `checkLHUpgradePath` to the beginning of `CheckUpgradePath`? so we don't need to pass the variable to the downstream methods.  

```go
	if !enableUpgradeVersionCheck {
		if semver.Compare(lhCurrentVersion, "v1.4.0") < 0 {
			return fmt.Errorf("failed to skip the upgrade check for %s to upgrade to %s, as it only supports the current version larger than v1.4.0", lhCurrentVersion, meta.Version)
		}
		logrus.Warnf(skipUpgradeVersionCheckMsg, lhCurrentVersion, meta.Version)
		eventRecorder.Eventf(&corev1.ObjectReference{Namespace: namespace, Name: "longhorn-upgrade"}, corev1.EventTypeWarning, constant.EventReasonUpgrade, skipUpgradeVersionCheckMsg, lhCurrentVersion, meta.Version)
		return nil
	}
```

File: upgrade/util/util.go (hunk 1 above)
Comment (mantissahz) [reply upgrade/util/util.go]: Moved codes
```golang
if !enableUpgradeVersionCheck {
	if semver.Compare(lhCurrentVersion, "v1.4.0") < 0 {
		return fmt.Errorf("failed to skip the upgrade check for %s to upgrade to %s, as it only supports the current version larger than v1.4.0", lhCurrentVersion, meta.Version)
	}
	logrus.Warnf(skipUpgradeVersionCheckMsg, lhCurrentVersion, meta.Version)
	eventRecorder.Eventf(&corev1.ObjectReference{Namespace: namespace, Name: "longhorn-upgrade"}, corev1.EventTypeWarning, constant.EventReasonUpgrade, skipUpgradeVersionCheckMsg, lhCurrentVersion, meta.Version)
	return nil
}
```
from checkLHUpgradePath to the beginning of CheckUpgradePath.
And removed passing variables to downstream methods.
Thanks.

**Comment by innobead** [comment]
There are a few things to change after the engine upgrade version enforcement was introduced, which increased `	CLIAPIMinVersion` to 8 from 3, which is the version used in 1.5, 1.4, and 1.3.

- Add back the VA creation in the upgrade path like what we did in 1.4 -> 1.5
- Add a new field `incompatible` to engineer image status and remove the original [`incompatible` state](https://github.com/longhorn/longhorn-manager/blob/2931e5b7bbc88849f67e9eb9cdab32b3fa531da4/k8s/pkg/apis/longhorn/v1beta2/engineimage.go#L10), which will make `CheckEngineImageReadiness` incorrectly check the image readiness.
- Update the code using the `EngineImageStateIncompatible1` to use the new status filed `incompatible`.

[/datastore/longhorn.go#L1746-L1761](https://github.com/longhorn/longhorn-manager/blob/2931e5b7bbc88849f67e9eb9cdab32b3fa531da4/datastore/longhorn.go#L1746-L1761)
```go
func (s *DataStore) CheckEngineImageReadiness(image string, nodes ...string) (isReady bool, err error) {
	if len(nodes) == 0 {
		return false, nil
	}
	if len(nodes) == 1 && nodes[0] == "" {
		return false, nil
	}
	ei, err := s.GetEngineImageRO(types.GetEngineImageChecksumName(image))
	if err != nil {
		return false, errors.Wrapf(err, "failed to get engine image %v", image)
	}
	if ei.Status.State != longhorn.EngineImageStateDeployed &&
		ei.Status.State != longhorn.EngineImageStateDeploying {
		logrus.Warnf("CheckEngineImageReadiness: engine image %v is in state %v", image, ei.Status.State)
		return false, nil
	}
```

@mantissahz after fixing this, update https://github.com/longhorn/longhorn/blob/master/enhancements/20230815-engine-upgrade-enforcement.md. Also, use separate commits to make the implementation clear if needed.

**Comment by mantissahz** [comment]
Tested with the latest implementation:

After upgrading v1.4.x volume attaching to a workload to 1.6.x, we can live/offline upgrade the volume engine to the master-head engine image.

One warning operation:
- After the upgrade, detach the v1.4.x volume and then it should not re-attach v1.4.x volume without upgrading the engine.

**Comment by innobead** [comment]
What's the reason for this warning operation?

File: k8s/pkg/apis/longhorn/v1beta2/engineimage.go [hunk 2]
```diff
@@ -5,10 +5,9 @@ import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 type EngineImageState string
 
 const (
-	EngineImageStateDeploying    = EngineImageState("deploying")
-	EngineImageStateDeployed     = EngineImageState("deployed")
-	EngineImageStateIncompatible = EngineImageState("incompatible")
-	EngineImageStateError        = EngineImageState("error")
```
Comment (innobead) [inline k8s/pkg/apis/longhorn/v1beta2/engineimage.go:11]: Let's don't remove it, but mark it as deprecated. 

Also, make an upgrade path for updating the `incompatible` field by `status::state::EngineImageStateIncompatible`.

**Review by innobead — APPROVED** [review]

**Pull request merged**